### DISCUSSION TITLE
trouble passing setting to typst-lsp

### DISCUSSION THREAD

**Discussion by Zjl37** [emacs-lsp/lsp-mode#4305] [Q&A]
Hello. I'm configuring my Emacs to use [nvarner/typst-lsp](https://github.com/nvarner/typst-lsp), an LSP for a (relatively new) markup typesetting language [Typst](https://typst.app/docs/). and I wish to disable the lsp's "compile on save" feature, that is, whenever I save a *.typ file, I do not want typst-lsp to compile it into a PDF document with the same name in that same directory. According to [this](https://github.com/nvarner/typst-lsp/discussions/158), I need to change the setting `exportPdf` to `never`. However I can't find a way to make this setting work in Emacs after hours of attempt. below is what I did:

Since Typst is not a supported language by lsp-mode, nor did typst-lsp mention in its readme the install steps for Emacs, I referred to [Adding support for languages](https://emacs-lsp.github.io/lsp-mode/page/adding-new-language/) page, and come up with the following setup:

```elisp
(add-to-list 'lsp-language-id-configuration '(typst-ts-mode . "typst"))
(lsp-register-client
 (make-lsp-client :new-connection (lsp-stdio-connection "typst-lsp")
                  :activation-fn (lsp-activate-on "typst")
                  :server-id 'typst-lsp))

(add-hook 'typst-ts-mode-hook 'lsp-deferred)
```

(in which `typst-ts-mode` is already properly configured)

And, If i understood the document correct, I need to use `lsp-defcustom`. Feeding the `package.json` file from inside the [vsix VScode extension file](https://github.com/nvarner/typst-lsp/releases/download/v0.12.1/typst-lsp-linux-x64.vsix) into [`lsp-generate-settings`](https://github.com/emacs-lsp/lsp-mode/blob/master/scripts/lsp-generate-settings.el), with the second argument being `'lsp-typst-lsp`, I got this elisp code:

```elisp
(lsp-defcustom lsp-typst-lsp-export-pdf "onSave"
  "The extension can export PDFs of your Typst files. This setting controls whether this feature is enabled and how often it runs."
  :type '(choice (:tag never onSave onPinnedMainSave onType onPinnedMainType))
  :group 'lsp-typst-lsp
  :package-version '(lsp-mode . "8.0.1")
  :lsp-path "typst-lsp.exportPdf")
```

among other five similar `lsp-defcustom` expressions. I appended them to my emacs init file, then added `(setq lsp-typst-lsp-export-pdf "never")`.

The LSP did work -- completions, error checks, ... all works with my exisiting configuration. But the `exportPdf` = `never` setting is not in effect. I still see pdf generated after saving .typ files. In contrast in VS Code or VSCodium, using the extension also provided by nvarner/typst-lsp, I can easily find the setting in the GUI, choose "never", and see the desired effect.

What's even stranger, when I tried to `M-x` `customize-variable` `RET` `lsp-typst-lsp-export-pdf`, nothing appears in the eazy customization interface, under the three buttons. Meanwhile an error message appears: `widget-choice-match: Symbol’s function definition is void: nil`

Where am I doing wrong? Can someone tell me the right way to set these options?

Many thanks🙏🙏 in advance.

### Extra information

Emacs version: `GNU Emacs 29.1 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.17.8)`

(vanilla, no doom or other configuration layer)

lsp-mode version: 20240116.1748, installed from melpa

`typst-lsp --version`: `Version: 0.12.1, commit 1928d88ff8bc3e178ad098146bec252751383a48 (Typst version 0.10.0)`

typst-ts-mode (I believe this is not relevant) is [meow_king/typst-ts-mode on sourcehut](https://git.sr.ht/~meow_king/typst-ts-mode), latest commit, setup as such:

```elisp
(use-package typst-ts-mode
  :straight (:type git :host sourcehut :repo "meow_king/typst-ts-mode"))
```

`typst --version`: `typst 0.10.0 (70ca0d25)`

**Comment by Zjl37** [comment]
Just managed to get trace logs of communication between lsp server and clients, both in Emacs and VSCodium. Searched string `exportPdf` in Codium's log, it only appeared once in the first initialize request, in here:
![图片](https://github.com/emacs-lsp/lsp-mode/assets/42609018/f0ae05b8-7370-4509-ad02-657b02912d13)
while in Emacs's log the same place was just `null`

**Reply by yyoncho** [reply]
You have to do something like that: https://github.com/emacs-lsp/lsp-mode/blob/master/clients/lsp-volar.el#L120 (without the ht-merge part).

**Reply by Zjl37** [reply]
Thank you for pointing me out! It worked :+1:  after I passed `:initialization-options (ht ("exportPdf" "never"))` to `make-lsp-client`, or passing `:initialization-options (lambda () (gethash "typst-lsp" (lsp-configuration-section "typst-lsp")))`

however it seems I couldn't find `:initialization-options` documented anywhere, which is a bit frustrating.

Still, the customization interface don't work for these `lsp-defcustom`-defined variables, and I wonder why. Though I can do `setq`,  `(setopt lsp-typst-lsp-export-pdf "never")` would cause error with this backtrace:

```elisp
Debugger entered--Lisp error: (void-function nil)
  nil((:tag :args (never onSave onPinnedMainSave onType onPinnedMainType)) "never")
  widget-choice-match((choice :args ((:tag :args (never onSave onPinnedMainSave onType onPinnedMainType)))) "never")
  setopt--set(lsp-typst-lsp-export-pdf "never")
  (progn (setopt--set 'lsp-typst-lsp-export-pdf "never"))
  elisp--eval-last-sexp(nil)
  eval-last-sexp(nil)
  funcall-interactively(eval-last-sexp nil)
  command-execute(eval-last-sexp)
```

**Reply by DrLoiseau** [reply]
Thanks so much ! I had been trying to do that for weeks... I finally won't be getting all those empty pdfs for my packages .typ files.
